File: 3 - Searching & Sorting/Binary Search/binarySearch.cpp
#include <iostream>
#include <vector>
using namespace std;

// Binary search in applicable only on sorted array
// T.C : O(logn)
int binarySearch(vector<int> *arr, int target)
{

    int start = 0, end = (*arr).size() - 1;
    int mid = start + (end - start) / 2;

    while (start <= end)
    {
        if ((*arr)[mid] == target)
            return mid;
        else if ((*arr)[mid] > target)
        {
            end = mid - 1; // right
        }
        else if ((*arr)[mid] < target)
        {
            start = mid + 1; // left
        }
        mid = start + (end - start) / 2;
    }
    return -1;
}

int main()
{

    vector<int> arr = {1, 2, 3, 4, 5, 6, 7, 8};
    int target = 8;

    int a = binarySearch(&arr, target);
    cout << a << endl;
}

File: 2 - Arrays/1D Arr/binarysearch.java
class binarysearch{

    int binary(int arr[],int key){
        int start=0;
        int end=arr.length-1;
        int mid=start+(end-start)/2;
        while(start<=end){
            if(arr[mid]==key){
                return mid;
            }
            //FOR RIGHT PART
            if(key>arr[mid]){
                start=mid+1;
            }
            //FOR left part
            else{
                end=mid-1;
            }
            mid=start+(end-start)/2;
        }
        return -1;
    }


    public static void main(String[] args) {
        int arr[]={2,4,6,8,9,13,45,56,78};
        binarysearch b1=new binarysearch();
        //b1.binary(arr,56);
        System.out.println(b1.binary(arr,56));
    }
}